satellite/metainfo/metainfo.go

```diff
@@ -1273,9 +1295,93 @@ func (endpoint *Endpoint) CommitSegment(ctx context.Context, req *pb.SegmentComm
+		return nil, status.Errorf(codes.InvalidArgument, "invalid number of upload results: wanted max %d got %d",
+			len(segmentID.OriginalOrderLimits), len(req.UploadResult))
+	}
+
+	pieces := make([]*pb.RemotePiece, len(req.UploadResult))
+	for i, result := range req.UploadResult {
+		pieces[i] = &pb.RemotePiece{
+			PieceNum: result.PieceNum,
+			NodeId:   result.NodeId,
+			Hash:     result.Hash,
+		}
+	}
+	remote := &pb.RemoteSegment{
+		Redundancy:   streamID.Redundancy,
+		RootPieceId:  segmentID.RootPieceId,
+		RemotePieces: pieces,
+	}
+
+	pointer := &pb.Pointer{
+		Type:        pb.Pointer_REMOTE,
+		Remote:      remote,
+		SegmentSize: req.SizeEncryptedData,
+
+		CreationDate:   streamID.CreationDate,
+		ExpirationDate: streamID.ExpirationDate,
+	}
+
+	if segmentID.Index == lastSegment {
+		pointer.Metadata = streamID.EncryptedMetadata
+	}
+
+	orderLimits := make([]*pb.OrderLimit, len(segmentID.OriginalOrderLimits))
+	for i, orderLimit := range segmentID.OriginalOrderLimits {
+		orderLimits[i] = orderLimit.Limit
+	}
+
+	err = endpoint.validatePointer(ctx, pointer, orderLimits)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, err.Error())
+	}
+
+	err = endpoint.filterValidPieces(ctx, pointer, orderLimits)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, err.Error())
+	}
+
+	path, err := CreatePath(ctx, keyInfo.ProjectID, int64(segmentID.Index), streamID.Bucket, streamID.EncryptedPath)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, err.Error())
+	}
+
+	exceeded, limit, err := endpoint.projectUsage.ExceedsStorageUsage(ctx, keyInfo.ProjectID)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, err.Error())
+	}
+	if exceeded {
+		endpoint.log.Sugar().Errorf("monthly project limits are %s of storage and bandwidth usage. This limit has been exceeded for storage for projectID %s.",
+			limit, keyInfo.ProjectID,
+		)
+		return nil, status.Errorf(codes.ResourceExhausted, "Exceeded Usage Limit")
+	}
+
+	inlineUsed, remoteUsed := calculateSpaceUsed(pointer)
+
+	// ToDo: Replace with hash & signature validation
+	// Ensure neither uplink or storage nodes are cheating on us
+	if pointer.Type == pb.Pointer_REMOTE {
+		//We cannot have more redundancy than total/min
+		if float64(remoteUsed) > (float64(pointer.SegmentSize)/float64(pointer.Remote.Redundancy.MinReq))*float64(pointer.Remote.Redundancy.Total) {
+			endpoint.log.Sugar().Debugf("data size mismatch, got segment: %d, pieces: %d, RS Min, Total: %d,%d", pointer.SegmentSize, remoteUsed, pointer.Remote.Redundancy.MinReq, pointer.Remote.Redundancy.Total)
+			return nil, status.Errorf(codes.InvalidArgument, "mismatched segment size and piece usage")
+		}
+	}
+
+	if err := endpoint.projectUsage.AddProjectStorageUsage(ctx, keyInfo.ProjectID, inlineUsed, remoteUsed); err != nil {
+		endpoint.log.Sugar().Errorf("Could not track new storage usage by project %v: %v", keyInfo.ProjectID, err)
```

Comment:
Why not use here a `log.Error` rather than the sugar version?

Reply:
the same answer as previously